src/github.com/getlantern/flashlight/client/balancer.go

```diff
@@ -7,40 +7,70 @@ import (
 	"github.com/getlantern/fronted"
 )
 
+// getBalancer waits for a message from client.balCh to arrive and then it
+// writes it back to client.balCh before returning it as a value. This way we
+// always have a balancer at client.balCh and, if we don't have one, it would
+// block until one arrives.
 func (client *Client) getBalancer() *balancer.Balancer {
+
+	// This balChMu will protect balCh and ensure it always have, at most, one
+	// element enqueued.
+	client.balChMu.Lock()
+	defer client.balChMu.Unlock()
+
 	bal := <-client.balCh
 	client.balCh <- bal
+
 	return bal
 }
 
+// initBalancer takes hosts from cfg.FrontedServers and cfg.ChainedServers and
+// it uses them to create a balancer. It also looks for the highest QOS dialer
+// available among the fronted servers.
 func (client *Client) initBalancer(cfg *ClientConfig) (*balancer.Balancer, fronted.Dialer) {
+	var highestQOSFrontedDialer fronted.Dialer
+
+	// The dialers slice must be large enough to handle all fronted and chained
+	// servers.
 	dialers := make([]*balancer.Dialer, 0, len(cfg.FrontedServers)+len(cfg.ChainedServers))
 
+	// Adding fronted servers.
 	log.Debugf("Adding %d domain fronted servers", len(cfg.FrontedServers))
-	var highestQOSFrontedDialer fronted.Dialer
 	highestQOS := math.MinInt32
 	for _, s := range cfg.FrontedServers {
+		// Getting a dialer for domain fronting and a dialer to dial to arbitrary
+		// addreses.
 		fd, dialer := s.dialer(cfg.MasqueradeSets)
 		dialers = append(dialers, dialer)
 		if dialer.QOS > highestQOS {
+			// If this dialer as a higher QOS than our current highestQOS, set it as
+			// the highestQOSFrontedDialer.
 			highestQOSFrontedDialer = fd
+			highestQOS = dialer.QOS
```

Comment:
We were missing a line that updated `highestQOS` to the greatest known value of `dialer.QOS` in order to actually get the highest QOS.